### PR TITLE
Generate GAS-compatible 32-bit ARM ASM files for libvpx on Linux.

### PR DESCRIPTION
**Pull request by dusxmt**
Bugfix for https://github.com/desktop-app/tg_owt/issues/48

Note: I wasn't able to test the full build because of the protobuf incompatibility in https://github.com/desktop-app/tg_owt/issues/43, but libvpx does indeed build now.

### CODE REVIEW

File: cmake/libvpx.cmake [hunk 1]
```diff
@@ -694,31 +701,68 @@ elseif (is_arm OR is_aarch64)
...
+                OUTPUT ${out_file_path}
+                COMMAND sh -c \"cat \'${in_file_path}\' | ${PERL_EXECUTABLE} \'${ads2gas}\' > \'${out_file_path}\'\"
+                DEPENDS ${in_file_path}
+            )
+
+            nice_target_sources(libvpx ${libvpx_arm32_asm_dir_obj}
+            PRIVATE
+                ${out_file}
+            )
+        endforeach()
```
Comment (ilya-fedin) [inline cmake/libvpx.cmake:757]: Probabluy the right way is to create a target_ads2gas_sources, just like target_yasm_source and give it .asm files

Comment (dusxmt) [reply cmake/libvpx.cmake:757]: Thanks for the suggestion, I'll try to re-factor it on the following weekend if I'll have time

File: cmake/libvpx.cmake [hunk 2]
```diff
@@ -69,7 +105,13 @@ else()
         )
     endif()
 
-    set(ASM_SUFFIX ".asm.S")
+endif()
+
+if (ads2gas)
+    find_package(Perl)
+    if (NOT PERL_FOUND)
+        message(FATAL_ERROR "Unable to find perl, needed by ${ads2gas} on 32-bit ARM systems when NEON SIMD is enabled.")
```
Comment (ilya-fedin) [inline cmake/libvpx.cmake:113]: why not just `find_package(Perl REQUIRED)`?

Comment (dusxmt) [reply cmake/libvpx.cmake:113]: I guess my lack of experience with cmake is showing :smile:  Certainly, we can do that.  I guess one advantage the current approach has is that it explains _why_ perl is needed.

File: cmake/libvpx.cmake (hunk 1 above)
Comment (dusxmt) [reply cmake/libvpx.cmake]: Finally found some free time, and agreed, it looks much better now.  (my experience with cmake is rather limited)

I put it into the libvpx.cmake file, since the ads2gas scripts are specific to libvpx, but if you feel like it should be in a separate file, I guess we could move it.  (it would probably need to be made more generic then, with parameters for a script file and such)

File: cmake/libvpx.cmake (hunk 2 above)
Comment (ilya-fedin) [reply cmake/libvpx.cmake]: i guess anyone who want to understand why it is needed can look into cmake files, cmake show the line what triggered the error AFAIR

File: cmake/libvpx.cmake [hunk 3]
```diff
@@ -59,6 +93,8 @@ else()
         list(APPEND include_directories
             ${libvpx_loc}/source/config/linux/arm-neon
         )
+        set(ads2gas ${libvpx_loc}/source/libvpx/build/make/ads2gas.pl)
```
Comment (ilya-fedin) [inline cmake/libvpx.cmake:96]: why is this needed? why just don't specify the path inside target_ads2gas_sources?

File: cmake/libvpx.cmake (hunk 2 above)
Comment (dusxmt) [reply cmake/libvpx.cmake]: Changed to `find_package(Perl REQUIRED)`.

File: cmake/libvpx.cmake (hunk 3 above)
Comment (dusxmt) [reply cmake/libvpx.cmake:96]: That's because there's actually more variants of the script, like e.g. for 32-bit Windows or Apple Devices; we don't currently use any of those.

Another reason is to allow the check for Perl to be conditional

File: cmake/libvpx.cmake (hunk 1 above)
Comment (ilya-fedin) [reply cmake/libvpx.cmake]: I think it's fine. Gladly, we have #55 also now. Thank you very much.

Comment (dusxmt) [reply cmake/libvpx.cmake]: You're welcome, I'm sorry it took so long.

File: cmake/libvpx.cmake (hunk 3 above)
Comment (ilya-fedin) [reply cmake/libvpx.cmake:96]: It feels too hackish to me, why don't do the check for the system inside the function?

You can also move find_package inside the function, look here: https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/cmake/generate_scheme.cmake

Comment (dusxmt) [reply cmake/libvpx.cmake:96]: Thanks, that's a good idea, will do!

File: cmake/libvpx.cmake [hunk 4]
```diff
@@ -9,6 +9,51 @@ set(include_directories
...
+        set(native_name ${gen_dst}/${entry}.S)
+
+        get_filename_component(native_name_loc ${native_name} DIRECTORY)
+        file(MAKE_DIRECTORY ${native_name_loc})
+
+        add_custom_command(
+        OUTPUT
+            ${native_name}
+        COMMAND
+            sh -c \"cat \'${ads_name}\' | ${PERL_EXECUTABLE} \'${ads2native}\' > \'${native_name}\'\"
```
Comment (ilya-fedin) [inline cmake/libvpx.cmake:44]: Well... looks weird 🤪

Comment (ilya-fedin) [reply cmake/libvpx.cmake:44]: Probably you should use [VERBATIM](https://cmake.org/cmake/help/latest/command/add_custom_command.html)

Comment (ilya-fedin) [reply cmake/libvpx.cmake:44]: Also, why cat is used instead of `<`?

Comment (dusxmt) [reply cmake/libvpx.cmake:44]: Is there a way to pipe standard input and standard output through an application in cmake?  That's the reason why I used the shell for it.

Comment (ilya-fedin) [reply cmake/libvpx.cmake:44]: I thought cmake passes the command to the shell, doesn't it?

Comment (dusxmt) [reply cmake/libvpx.cmake:44]: I'm not sure, I guess I can try it and see if it works

Comment (dusxmt) [reply cmake/libvpx.cmake]: It did indeed work, thanks!  With your help, I might even learn cmake :smile:

**Comment by ilya-fedin** [comment]
Looks beautiful for me now! I hope @john-preston will be able to look at this soon.

**Comment by john-preston** [comment]
Thanks!

**Comment by ilya-fedin** [comment]
Tried to re-enable NEON on snapcraft builds, armhf still doesn' build with NEON :(
photo_2021-05-26_15-58-53

**Comment by dusxmt** [comment]
Thanks for letting me know!  Could you try it with https://github.com/desktop-app/tg_owt/pull/60?

**Comment by ilya-fedin** [comment]
Well, it's not that simple. Snapcraft infrastructure builds only source that will be pushed to end users via edge channels. I.e. that would require to change tg_owt URL in snapcraft.yaml in tdesktop repo. This probably is not a good idea...

**Comment by dusxmt** [comment]
I see.  Well, the patch should fix it, there were two .c files which seemed like they were mutually exclusive with .asm files, but it turned out that they weren't.

The remaining files in the list were indeed mutually exclusive, I checked them each with grep and saw that the functions they exported had the same names as functions in the .asm files - would result in a redefined symbol error if they were included in the 32-bit build.
